# Week6_Web APIs

An API, or Application Program Interface, allows one program to talk to another program. Many websites or services provide an API so you can query for information or download datasets in an automated way. For example, many cities provide public transit APIs, such as GTFS Realtime or Google Maps Transit API, to share live bus and train locations, estimated arrival times, and delays. A transit app can use these APIs to show real-time vehicle positions on a map and provide updated arrival predictions for passengers.


In this lab, we will introduce the Geocoding APIs and Distance Matrix applications provided by the Google Map API Services.
<p><a id="table"> </a></p>
<h2 id="Table-of-Contents">Table of Contents<a class="anchor-link" href="#Table-of-Contents"></a></h2>
<ol>
<h3>1 <a href="#googlemapapi">Google Map API</a></h3>
1.1 <a href="#preparation">Preparation</a><br/>
1.2 <a href="#geocodingapi">Geocoding API</a><br/>  
1.3 <a href="#distancematrixapi">Distance matrix API</a><br/>
</ol>

<p><a id="googlemapapi"> </a></p>
<h2 id="googlemapapi">1 <a href="#table">Google Map API</a><a class="anchor-link" href="#googlemapapi">¶</a></h2>
  

<p><a id="preparation"> </a></p>
<h3 id="preparation">1.1 <a href="#table">Preparation</a><a class="anchor-link" href="#preparation">¶</a></h3>

To use the Google API, we need to prepare two things: Google API account and Google API library for Python.

- First, we need to create a Google API account with an associated `API Key`. The key is used to identify you and enforce usage limits so that you do not overwhelm the servers. I provide a key from my own account for **`instruction purpose only`**. However, Please create your own project and API key for your own projects. The instructions for creating projects and API key is available here: https://developers.google.com/maps/gmp-get-started#api-key. According to the current Google Cloud policy, new accounts will receive 300 dolloar in credits, and there is a 200 dollar free quota per month. If you exceed this limit, you will be charged.

- Second, the Python library for Google API is called `googlemaps`. To install the library, we type in `!pip install googlemaps` in the Jupyter Notebook.  

The googlemaps library provides us a list of applications using Google Map APIs: 
- Directions API
- Distance Matrix API
- Elevation API
- Geocoding API
- Geolocation API
- Time Zone API
- Roads API
- Places API
- Maps Static API
    
Check out the link of `googlemaps` library for more API services: https://github.com/googlemaps/google-maps-services-python 

Let us firstly import the library `googlemaps` - the Python Client for Google Maps Services.

In [11]:
# run this code to install Google Map API
!pip install googlemaps

Then we create a Client object using API key before Geocoding:

In [23]:
# import the googlemaps library

import googlemaps

# create a Client object
gmaps = googlemaps.Client(key='AIzaSyDKp7QO8kJvVIG1Be5WG4qBjgYzea_qKr4')

In [24]:
# import other libraries 

import warnings
# Ignore all warnings
warnings.filterwarnings('ignore')

import pandas as pd

<p><a id="geocodingapi"> </a></p>
<h3 id="geocodingapi">1.2 <a href="#table">Geocoding API</a><a class="anchor-link" href="#geocodingapi">¶</a></h3>

 
In this section, we will introduce two basic applications

- the **Geocoding API** which returns the corresponding longitude and latitude in terms of the address,
- and the **Distance Matrix API** which calculates the real-time travel distance/time from one location to another in terms of travel modes.

To begin with, **go to the Lab folder and load the COVID-19 NYC Vaccination sites dataset (COVID19_Vaccination.xlsx)**. This data is a subset of data aggregated by GISCorps volunteers and is sourced from public information shared by health departments, local governments, and healthcare providers. The entire dataset can be downloaded at https://covid-19-giscorps.hub.arcgis.com/search?collection=Dataset

The dataset stores the vaccination site names and addresses in the five boroughs of the New York City. However, it is non-spatial dataset due to the lack of geographic coordinates. What we will do in the first step is to convert the vaccination addresses into longitude and latitude, so that we can plot the testing sites on maps.  

In [75]:
# select and display the top 10 rows
df_NYC = pd.read_excel("COVID19_Vaccination.xlsx")
df_NYC = df_NYC.iloc[:10]
df_NYC

,SiteID,Name,Address1,Address2
0,2664,"Gotham Health, Morrisania",1225 Gerard Avenue,"Bronx, New York 10452"
1,2843,Metropolitan,1901 First Avenue,"New York, NY 10029"
2,3050,StatCare in Bronx (Bartow Mall),2063A Bartow Avenue Bronx,NY 10475
3,4271,StatCare Bronx (174th St.) Urgent Care,932 East 174th Street Bronx,NY 10460
4,4913,Statcare COVID-19 Testing Tents - Brooklyn,341 Eastern Parkway,"Brooklyn, NY 11216"
5,4915,Statcare COVID-19 Testing Tents - Astoria,"37-15 23rd Ave, Astoria",NY 11105
6,4916,Statcare COVID-19 Testing Tents - Jackson Heights,"80-10 Northern Blvd, Jackson Heights",NY 11372
7,21203,Northwell Health-GoHealth Urgent Care,4316 Amboy Road,"Staten Island, NY 10312"
8,21211,Northwell Health-GoHealth Urgent Care,"102-29 Queens Blvd, Forest Hills",NY 11375
9,21212,Northwell Health-GoHealth Urgent Care,199 Amsterdam Ave,"New York, NY 10023"


We find that the DataFrame contains two address columns. `Address1` stores the street number and `Address 2` stores the borough, city, and zipcode. In geocoding, **the more completed the address, the more accurate the latitude and longitude you can obtain**. For example, if you perform geocoding using only the information from Address 1, the Google Maps API is likely to find multiple addresses with the same street number located in different cities. Therefore, we need to combine the addresses from Address 1 and Address 2 to ensure more accurate geocoding results.

In [77]:
# Combining two columns into one
df_NYC['combined_address'] = df_NYC['Address1'] + ', ' + df_NYC['Address2']
df_NYC.head()

,SiteID,Name,Address1,Address2,combined_address
0,2664,"Gotham Health, Morrisania",1225 Gerard Avenue,"Bronx, New York 10452","1225 Gerard Avenue, Bronx, New York 10452"
1,2843,Metropolitan,1901 First Avenue,"New York, NY 10029","1901 First Avenue, New York, NY 10029"
2,3050,StatCare in Bronx (Bartow Mall),2063A Bartow Avenue Bronx,NY 10475,"2063A Bartow Avenue Bronx, NY 10475"
3,4271,StatCare Bronx (174th St.) Urgent Care,932 East 174th Street Bronx,NY 10460,"932 East 174th Street Bronx, NY 10460"
4,4913,Statcare COVID-19 Testing Tents - Brooklyn,341 Eastern Parkway,"Brooklyn, NY 11216","341 Eastern Parkway, Brooklyn, NY 11216"



#### 1. The function `.geocode()`
Geocoding is the process of converting addresses (like a street address) into geographic coordinates (latitude and longitude). See the following example: 

![Geocoding Image](https://github.com/wenzhengli-etal/UrbanDataScience_4680_5680/blob/f7392c2eff92aa4883edaa9e03f620d360c2baca/img/geocode.jpg?raw=true)

- Note that the Google Map Geocoding API returns the World Geodetic System (WGS-1984).  

Now go back to the vaccination DataFrame, and convert the actual address to longitude and latitude using the `.geocode()`, a function of the googlemaps library, which send the Geocoding query to the Google Map server and gets the response as WGS84 longitude and latitude.

In [78]:
# check the address of the first row
df_NYC.loc[0, 'combined_address']

# Geocoding the first address
geocode_result = gmaps.geocode(df_NYC.loc[0,'combined_address'])
geocode_result

[{'address_components': [{'long_name': '1225',
    'short_name': '1225',
    'types': ['street_number']},
   {'long_name': 'Gerard Avenue',
    'short_name': 'Gerard Ave',
    'types': ['route']},
   {'long_name': 'Concourse',
    'short_name': 'Concourse',
    'types': ['neighborhood', 'political']},
   {'long_name': 'The Bronx',
    'short_name': 'Bronx',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': 'Bronx County',
    'short_name': 'Bronx County',
    'types': ['administrative_area_level_2', 'political']},
   {'long_name': 'New York',
    'short_name': 'NY',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'United States',
    'short_name': 'US',
    'types': ['country', 'political']},
   {'long_name': '10452', 'short_name': '10452', 'types': ['postal_code']},
   {'long_name': '8001',
    'short_name': '8001',
    'types': ['postal_code_suffix']}],
  'formatted_address': '1225 Gerard Ave, Bronx, NY 10452, USA',
  'ge

In [32]:
# # convert to a well-structured json string. 
# import json

# # Pretty-printing the JSON data
# pretty_data = json.dumps(geocode_result, indent=4)
# print(pretty_data)

In [79]:
# how many addresses it returns?
len(geocode_result)

1

The geocode_result is a list containing one dictionary elements, starts with "{'address_components':...".

Let us look at component it has: 

1. `address_components[]` is an array that categorizes an input address into separated components and parses each of them based on `long name`, `short name` and `type`. `long_name` is the full text description or name of the address component; `short name` is the abbreviated of the long name; `types:[]` is the type of long name and short name address component,  for example, '1225' is a `['street_number']`, 'United States' is identified as a `['country']`. Here are the type of addresses:    
    - Street Number: 1225
    - Route: Gerard Avenue
    - Neighborhood: Concourse
    - Sublocality: The Bronx (a borough of New York City)
    - Administrative Area Level 2: Bronx County
    - Administrative Area Level 1: New York (state)
    - Country: United States
    - Postal Code: 10452
    - Postal Code Suffix: 8001


3. `Formatted Address` is the next element (key) in dictionary which contains the human-readable address of the geocode location. The address is 1225 Gerard Ave, Bronx, NY 10452, USA - a complete address in a user-friendly format.

4. `geometry` contains two important information:
    - `location`: the geocoded latitude, longitude value. For normal address lookups, this field is typically the most important.
    - `location_type` the additional data about the specified location.

      
5. `Place ID`: ChIJpYh8ajr0wokRAbrfwxGLJyU - A unique identifier for this location in the Google Places database.

6. `Types`: The location is classified as premise, which generally refers to a building or property.

for more explanation: https://developers.google.com/maps/documentation/geocoding/overview#reverse-status


Now, let us obtain the returned address and the corresponding longitude and latitude: 

In [39]:
print(geocode_result[0]['formatted_address'])
print(geocode_result[0]['geometry']['location']['lng'])
print(geocode_result[0]['geometry']['location']['lat'])

1225 Gerard Ave, Bronx, NY 10452, USA
-73.9202294
40.8362867


#### 3. Test Geocoding API in a for-loop

We can test API for all addresses in a for-loop and save all results into a list for future use. Notice that sometimes the API can return multiple results (due to similar address), especially when the address is not well-specified. So, in collecting the full list of coordinates, one concern is the number of results returned for each address. We can use the `len(geocode_result)` during the loop to check.

In [61]:
# loop over the first 10 records.
result_list = []

for i in df_NYC["combined_address"].tolist():
    temp = gmaps.geocode(i)
    result_list.append(temp)
    print(len(temp), end = ", ")

1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 

In [69]:
# check the outstanding records
result_list[-2]

[{'address_components': [{'long_name': '102-29',
    'short_name': '102-29',
    'types': ['street_number']},
   {'long_name': 'Queens Boulevard',
    'short_name': 'Queens Blvd',
    'types': ['route']},
   {'long_name': 'Forest Hills',
    'short_name': 'Forest Hills',
    'types': ['neighborhood', 'political']},
   {'long_name': 'Queens',
    'short_name': 'Queens',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': 'Queens County',
    'short_name': 'Queens County',
    'types': ['administrative_area_level_2', 'political']},
   {'long_name': 'New York',
    'short_name': 'NY',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'United States',
    'short_name': 'US',
    'types': ['country', 'political']},
   {'long_name': '11375', 'short_name': '11375', 'types': ['postal_code']},
   {'long_name': '3258',
    'short_name': '3258',
    'types': ['postal_code_suffix']}],
  'formatted_address': '102-29 Queens Blvd, Forest Hill

In [70]:
# loop over the first 10 records, and if multiple elements, select the first one.
result_list = []

for i in df_NYC["combined_address"].tolist():
    temp = gmaps.geocode(i)

    if len(temp) != 1:
        temp = [temp[0]]
    result_list.append(temp)
    print(len(temp), end = ", ")

1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

Each address returns the single geocode address for all the 10 addresses, meaning our geocoding result should be accurate! Now, we can extract the useful information and save them in appropriate form (DataFrame). To do that, we firstly take the geographic coordinates from the result for each address by looping over the result list and then, combine the lists of `address`, `refined address`, `longitude`, and `latitude` into a DataFrame. 

In [50]:
# create two lists for latitude and longitude
lat_list = []
lng_list = []
refined_list = []

for r in result_list:
    refined_address = r[0]['formatted_address']
    refined_list.append(refined_address)
    
    lat = r[0]['geometry']['location']['lat']
    lat_list.append(lat)
    
    lng = r[0]['geometry']['location']['lng']
    lng_list.append(lng) 
    

In [52]:
# combine lists into a DataFrame
df = pd.DataFrame({
    'address': df_NYC['combined_address' ].tolist(),
    'refined_address': refined_list,
    'longitude':lng_list,
    'latitude':lat_list})
df

,address,refined_address,longitude,latitude
0,"1225 Gerard Avenue, Bronx, New York 10452","1225 Gerard Ave, Bronx, NY 10452, USA",-73.920229,40.836287
1,"1901 First Avenue, New York, NY 10029","1901 1st Ave., New York, NY 10029, USA",-73.945029,40.785233
2,"2063A Bartow Avenue Bronx, NY 10475","2063A Bartow Ave, Bronx, NY 10475, USA",-73.828381,40.870179
3,"932 East 174th Street Bronx, NY 10460","932 E 174th St, Bronx, NY 10460, USA",-73.887255,40.837020
4,"341 Eastern Parkway, Brooklyn, NY 11216","341 Eastern Pkwy, Brooklyn, NY 11216, USA",-73.957571,40.670892
5,"37-15 23rd Ave, Astoria, NY 11105","37-15 23rd Ave, Astoria, NY 11105, USA",-73.908927,40.771901
6,"80-10 Northern Blvd, Jackson Heights, NY 11372","80-10 Northern Blvd, Jackson Heights, NY 11372...",-73.886899,40.755178
7,"4316 Amboy Road, Staten Island, NY 10312","4316 Amboy Rd, Staten Island, NY 10312, USA",-74.159295,40.545385
8,"102-29 Queens Blvd, Forest Hills, NY 11375","102-29 Queens Blvd, Forest Hills, NY 11375, USA",-73.850369,40.725683
9,"199 Amsterdam Ave, New York, NY 10023","199 Amsterdam Ave, New York, NY 10023, USA",-73.982804,40.776423


<p><a id="distancematrixapi"> </a></p>
<h3 id="distancematrixapi">1.3 <a href="#table">Distance matrix API</a><a class="anchor-link" href="#distancematrixapi">¶</a></h3>


The Distance Matrix API is a service that provides travel distance and time for a matrix (set) of origins and destinations. The API returns information based on the recommended route between start and end points, as calculated by the Google Maps API, and consists of rows containing duration and distance values for each pair. This API is extremely convenient and scalable for batch requests determining aggregated metrics of routes (it does not return detailed navigation information, use the directions API for that).

Similar to the Google Map Geocoding API, we can use the function called `.distance_matrix()` in the `googlemaps` library to easily get the travel distance and journal duration. Check the technical manual for the Google Map Distance Matrix here: https://developers.google.com/maps/documentation/distance-matrix/overview

Let us take an example by calculating the travel distance and journey duration among several large cities near Ithaca. The first step is to define a list of the five cities and then create a list `lst_city` that saves all the possible city pairs among the five. The function `combinations()` in the `itertools` package help us to do that.     

In [89]:
from itertools import combinations
lst_city = ["New York City", "Ithaca", "Syracuse", "Toronto"]
city_pair = list(combinations(lst_city, 2)) # Nonrepeating city pairs (all possible combinations)
city_pair

[('New York City', 'Ithaca'),
 ('New York City', 'Syracuse'),
 ('New York City', 'Toronto'),
 ('Ithaca', 'Syracuse'),
 ('Ithaca', 'Toronto'),
 ('Syracuse', 'Toronto')]

Notice that the result returns a `list of tuples`. A tuple is an ordered, immutable collection of elements in Python. Each tuple contains a pair of city names. 

Next, let us calculate the distance and driving duration for the first city pair: `New York City`&mdash;`Los Angeles` using `.distance_matrix()`. 

The arguments for the `.distance_matrix()` can be found here: https://github.com/googlemaps/google-maps-services-python/blob/master/googlemaps/distance_matrix.py. Refer to it when you want to specify the transportation mode and compute routes which adhere to certain restrictions, such as avoiding specific road types or object characteristics.

In [90]:
# calculate the travel time and distance based on real-time traffic

from datetime import datetime
now = datetime.now()

my_dist = gmaps.distance_matrix(city_pair[0][0],city_pair[0][1], 
                                mode="driving",
                                departure_time = now,
                                traffic_model="best_guess"
                                )
  
# Printing the result 
my_dist 

{'destination_addresses': ['Ithaca, NY, USA'],
 'origin_addresses': ['New York, NY, USA'],
 'rows': [{'elements': [{'distance': {'text': '395 km', 'value': 395009},
     'duration': {'text': '4 hours 13 mins', 'value': 15150},
     'duration_in_traffic': {'text': '4 hours 13 mins', 'value': 15204},
     'status': 'OK'}]}],
 'status': 'OK'}

In [94]:
# calculate the travel time and distance on Feb 28 at 5am. 

from datetime import datetime
departure_time = datetime(2025, 2, 28, 5, 0, 0)

my_dist = gmaps.distance_matrix(city_pair[0][0],city_pair[0][1], 
                                mode="driving",
                                departure_time = departure_time,
                                traffic_model="best_guess"
                                )
  
# Printing the result 
my_dist 

{'destination_addresses': ['Ithaca, NY, USA'],
 'origin_addresses': ['New York, NY, USA'],
 'rows': [{'elements': [{'distance': {'text': '358 km', 'value': 358500},
     'duration': {'text': '3 hours 54 mins', 'value': 14045},
     'duration_in_traffic': {'text': '3 hours 53 mins', 'value': 14009},
     'status': 'OK'}]}],
 'status': 'OK'}

Similar to Geocoding API, the result returns a large dictionary that contains the destination and origin address, the distance (in kilometers and meters) and the driving duration (in hours and seconds). 

Let us extract the travel distance and duration from the distance_matrix result:

In [95]:
# obtain the destination
print(my_dist["destination_addresses"][0]) 

# obtain the driving distance (in km)
print(my_dist["rows"][0]['elements'][0]['distance']['text'])

# obtain the driving duration (in days and hours)
print(my_dist["rows"][0]['elements'][0]['duration']['text'])

Ithaca, NY, USA
358 km
3 hours 54 mins


Finally, we can use a nested for-loop structure to get the distance and journey duration for all city pairs by the travel modes of driving and bicycling, and save the returned lists in a DataFrame:

In [96]:
# create an empty list to save the results 
origin = []
destination = []


driving_distance = []
driving_duration = []

bicycling_distance = []
bicycling_duration = []

for tm in ["driving", "bicycling"]:
    for cp in city_pair:

        result = gmaps.distance_matrix(cp[0],cp[1], mode=tm)   # distance matrix result
               
        if tm == "driving":
            origin.append(result["origin_addresses"][0]) # get the origin from dist
            destination.append(result["destination_addresses"][0]) # get the destination from dist
            driving_distance.append(result["rows"][0]['elements'][0]['distance']['text'])  # get the distance from dist
            driving_duration.append(result["rows"][0]['elements'][0]['duration']['text'])  # get the duration from dist
        
        else: 
            bicycling_distance.append(result["rows"][0]['elements'][0]['distance']['text'])  # save into the list
            bicycling_duration.append(result["rows"][0]['elements'][0]['duration']['text']) 

In [97]:
# combine lists into a DataFrame
df = pd.DataFrame({
    'origin': origin,
    'destination': destination,
    'driving distance':driving_distance,
    'driving duration':driving_duration,
    'bicycling distance':bicycling_distance,
    'bicycling duration':bicycling_duration})
df

,origin,destination,driving distance,driving duration,bicycling distance,bicycling duration
0,"New York, NY, USA","Ithaca, NY, USA",358 km,3 hours 54 mins,395 km,22 hours 52 mins
1,"New York, NY, USA","Syracuse, NY, USA",397 km,4 hours 1 min,569 km,1 day 7 hours
2,"New York, NY, USA","Toronto, ON, Canada",758 km,7 hours 53 mins,997 km,2 days 5 hours
3,"Ithaca, NY, USA","Syracuse, NY, USA",90.5 km,1 hour 7 mins,119 km,6 hours 14 mins
4,"Ithaca, NY, USA","Toronto, ON, Canada",397 km,4 hours 16 mins,433 km,22 hours 33 mins
5,"Syracuse, NY, USA","Toronto, ON, Canada",393 km,3 hours 59 mins,428 km,22 hours 24 mins
